src/coreclr/jit/lower.cpp

```diff
@@ -5033,6 +5033,28 @@ bool Lowering::TryCreateAddrMode(GenTree* addr, bool isContainable, var_types ta
         }
     }
 
+#ifdef TARGET_ARM64
+    // Check if we can "contain" LEA(BFIZ) in order to extend 32bit index to 64bit as part of load/store.
+    if (index != nullptr && index->OperIs(GT_BFIZ) && index->gtGetOp1()->OperIs(GT_CAST) &&
+        index->gtGetOp2()->IsCnsIntOrI() && varTypeIsIntegral(targetType))
+    {
+        // BFIZ node is a binary op where op1 is GT_CAST and op2 is GT_CNS_INT
+        GenTreeCast* cast = index->gtGetOp1()->AsCast();
+        assert(cast->isContained());
+
+        const unsigned shiftBy = (unsigned)index->gtGetOp2()->AsIntCon()->IconValue();
+
+        // 'scale' and 'offset' have to be unset since we're going to use [base + index * SXTW/UXTW scale] form
+        // where there is no room for additional offsets/scales on ARM64. 'shiftBy' has to match target's width.
+        if (varTypeIsInt(cast->CastFromType()) && varTypeIsLong(cast->CastToType()) &&
+            (genTypeSize(targetType) == (1U << shiftBy)) && (scale == 1) && (offset == 0))
```

Comment:
It would be cheaper to rewrite these checks as the equivalent `cast->CastOp()->TypeIs(TYP_INT) && cast->TypeIs(TYP_LONG)`.

Reply:
@SingleAccretion what do you mean cheaper - it will skip cases when e.g. cast->CastToType() is TYP_ULONG

Reply:
> it will skip cases when e.g. cast->CastToType() is TYP_ULONG won't it?

I don't believe so. If `CastToType` is `TYP_LONG/TYP_ULONG`, the type of the cast node itself is always `TYP_LONG`.

Reply:
I put
```
            assert(cast->CastFromType() == TYP_INT);
            assert(cast->CastToType() == TYP_LONG);
```
and SPMI failed immediately

Reply:
On the first or the second assert?

Failure on the first assert means something is creating `TYP_UINT`s in IR (this is a bug).
The second assert failure is expected, `CastToType` can be `TYP_ULONG`.

What I asserted would be equivalent to this: `assert(!varTypeIsLong(cast->CastToType()) || !cast->TypeIs(TYP_LONG))`.

Reply:
@SingleAccretion it's the second assert - so I guess the check you suggest `cast->CastOp()->TypeIs(TYP_INT) && cast->TypeIs(TYP_LONG)` will skip cases with it

Reply:
I do not see how that is the case. The current check is `CastToType() == TYP_LONG || CastToType() == TYP_ULONG`.

If the `CastToType` is one of those two, the cast node's type (i .e `cast->TypeGet()`, which is what `TypeIs` tests) will be `TYP_LONG`, the reverse is also true. This follows from the fact that nodes do not (and cannot) have `TYP_ULONG`.

FWIW, I have built a cast checker in the past that asserted these invariants, and it did not fail (for them at least, there were a lot of other asserts in it that did fail 😄), so I am pretty confident in this.

Reply:
@SingleAccretion ah, I am blind 🤦‍♂️, I didn't notice that you check `cast->TypeIs(TYP_LONG)`, I thought you was proposing `cast->CastToType() == TYP_LONG` 😄